tests/host_tools/fcmetrics.py

```diff
@@ -363,3 +365,135 @@ def validate_per_device_metrics(self, fc_metrics):
+    for group, keys in fc_metrics.items():
+        if group == "utc_timestamp_ms":
+            continue
+        for key, value in keys.items():
+            full_key = f"{group}.{key}"
+            # values are 0 when:
+            # - there is no update
+            # - device is not used
+            # - SharedIncMetric reset to 0 on flush so if
+            #   there is no change metric the values remain 0.
+            # We can save the amount of bytes we export to
+            # CloudWatch in these cases.
+            # however it is difficult to differentiate if a 0
+            # should be skipped or upload because it could be
+            # an expected value in some cases so we upload
+            # all the metrics even if data is 0.
+            if full_key in skip:
+                continue
+            unit = get_emf_unit_for_fc_metrics(key)
+            metrics.put_metric(f"fc_metrics.{full_key}", value, unit=unit)
+            # print(f"fc_metrics.{full_key} {value}, unit={unit}")
+            if full_key in metrics_to_export_once:
+                skip.add(full_key)
+
+
+class FCMetricsMonitor(Thread):
+    """
+    read Firecracker metrics from the microvm every 60secs and
+    uploads the metrics to CW.
+    We do this as a daemon thread every 60sec, instead of collecting
+    all metrics together in the end, to retain timestamp of the metrics.
+    """
+
+    def __init__(self, vm, metrics_logger, timer=60):
+        Thread.__init__(self, daemon=True)
+        self.vm = vm
+        self.timer = timer
+
+        self.metrics_index = 0
+        self.running = True
+
+        self.metrics_logger = metrics_logger
+
+    def _flush_metrics(self):
+        """
+        Since vm.flush_metrics provides only the latest metrics,
+        we call vm.get_all_metrics() instead to be able to collect
+        and upload all metrics data emitted by the microvm.
+        This utility function is created to keep common code in one
+        place and is called every 60sec once the daemon starts and then
+        once when the daemon stops.
+        """
+        all_metrics = self.vm.get_all_metrics()
+        for metrics in all_metrics[self.metrics_index :]:
+            flush_fc_metrics_to_cw(metrics, self.metrics_logger)
+            self.metrics_index += 1
+
+    def stop(self):
+        """
+        Stop the daemon gracefully.
+        Since we depend on the vm to provide the metrics,
+        this method should be called just before killing the vm.
+        We collect final metrics here in stop instead of letting it
+        be collected from the "run" method because, "run" could be
+        in sleep when stop is called and once it wakes out of sleep
+        the "vm" might not be avaiable to provide the metrics.
+        """
+        self.running = False
+        # wait for the running thread to finish
+        # this should also avoid any race condition leading to
+        # uploading the same metrics twice
+        time.sleep(1)
```

Comment:
Would it make sense to wait for the thread here instead of waiting? `self.join()`

Reply:
yeah that looks much better, thanks!